src/ai/defaultai.cc

```diff
@@ -3341,70 +3344,219 @@ void DefaultAI::diplomacy_actions(const Time& gametime) {
 	const bool me_def = me->is_defeated();
 	const bool me_alone = player_statistics.get_is_alone(mypn);
 
-	// if we are defeated or the last one in a team leave team
+	// TODO(tothxa): Check all uses whether it's safe to add an invalid value to the enum
+	constexpr Widelands::DiplomacyAction kNoAction =
+	   static_cast<Widelands::DiplomacyAction>(std::numeric_limits<uint8_t>::max());
+	constexpr int32_t kNoScore = std::numeric_limits<int32_t>::min();
+
+	int32_t plan_priority = 0;
+	Widelands::DiplomacyAction planned_action = kNoAction;
+	Widelands::PlayerNumber planned_opn = 0;
+	int32_t planned_other_score = kNoScore;
+	std::string planned_log_append_text;
+
+	// If we are defeated or the last one in a team, then leave team, but check pending requests
+	// first.
+	// If alone, we may accept requests to join and cancel leaving.
+	// If defeated, just clean up by rejecting everything.
 	if (me->team_number() != 0 && (me_alone || me_def)) {
-		game().send_player_diplomacy(mypn, Widelands::DiplomacyAction::kLeaveTeam, 0 /* ignored */);
-		verb_log_dbg_time(gametime, "AI Diplomacy: Player(%d) leaves team (%d) as %s.\n",
-		                  static_cast<unsigned int>(mypn), static_cast<unsigned int>(mytn),
-		                  me_alone ? "last one" : "defeated");
+		planned_action = Widelands::DiplomacyAction::kLeaveTeam;
+		plan_priority = me_alone ? 0 : std::numeric_limits<int32_t>::max();
+		if (g_verbose) {
+			planned_log_append_text = me_alone ? " as last one" : " as defeated";
+			/* verb_ */ log_dbg_time(gametime,
+			                         "AI Diplomacy: Player(%d) plans to leave team with priority %d",
+			                         static_cast<unsigned int>(mypn), plan_priority);
+		}
 	}
 
 	const int32_t my_team_score =
 	   me_alone ? 0 : player_statistics.get_team_average_score(mytn, mypn);
 	const std::string myts_s = me_alone ? "" : format(" with team score %d", my_team_score);
 
-	bool request_accepted = false;
+	// Check desirability of current team
+	if (planned_action == kNoAction && !me_alone && my_team_score < 0 &&
+	    (my_team_score < -10 || RNG::static_rand(8) == 0)) {
+		planned_action = Widelands::DiplomacyAction::kLeaveTeam;
+		plan_priority = -my_team_score;
+		if (g_verbose) {
+			planned_log_append_text = myts_s;
+			/* verb_ */ log_dbg_time(gametime,
+			                         "AI Diplomacy: Player(%d) plans to leave team with priority %d",
+			                         static_cast<unsigned int>(mypn), plan_priority);
+		}
+	}
+
+	// Check for undesirable teammates
+	if (planned_action == kNoAction && player_statistics.get_worst_ally_score() < -15) {
+		const int32_t team_vs_worst = my_team_score + player_statistics.get_worst_ally_score();
```

Comment:
wouldn't this be ally score * 2 for a team with only 2 players. If yes we would leave earlier as before, wouldn't we

Reply:
I missed this one. But if we only have one bad ally, then the previous teamscore check also triggers earlier than before (-10 vs. -15), but I thought that makes sense. I'll add a check for this case to not overpenalise single bad ally.

Reply:
Actually this didn't affect when we leave single ally, only the priority for the plan to leave. The whole block depends on the old check of `diploscore < -15`, and neither of the inner conditions are met then, so the `else` branch is always executed for single ally.